sys/include/net/gnrc/ipv6/nib/conf.h

```diff
@@ -176,6 +176,17 @@ extern "C" {
 #define CONFIG_GNRC_IPV6_NIB_QUEUE_PKT                1
 #endif
 
+#if CONFIG_GNRC_IPV6_NIB_QUEUE_PKT
+/**
+ * @brief    queue capacity for the packets waiting for address resolution,
+ *           per neighbor. SHOULD always be smaller than @ref CONFIG_GNRC_IPV6_NIB_NUMOF
```

Comment:
Because if it's >= `CONFIG_GNRC_IPV6_NIB_NUMOF`, then even for a single neighbor, you can't even get to drop old packets from it's queue because you can't allocate a new one in the first place. With multiple neighbors, this can happen anyway, so this is why I went for a `SHOULD` instead of `MUST`.

Reply:
Ah
```C
#if IS_ACTIVE(CONFIG_GNRC_IPV6_NIB_QUEUE_PKT)
static gnrc_pktqueue_t _queue_pool[CONFIG_GNRC_IPV6_NIB_NUMOF];
#endif  /* CONFIG_GNRC_IPV6_NIB_QUEUE_PKT */
```

That is not just an array of queues, but rather every packet to be queued has to get a slot here.

Reply:
If `CONFIG_GNRC_IPV6_NIB_QUEUE_PKT_CAP` is the queue capacity per neighbor.
I would define this as 1 by default and change the upper code snipped to:

```C
#if IS_ACTIVE(CONFIG_GNRC_IPV6_NIB_QUEUE_PKT)
static gnrc_pktqueue_t _queue_pool[CONFIG_GNRC_IPV6_NIB_NUMOF * CONFIG_GNRC_IPV6_NIB_QUEUE_PKT_CAP];
#endif  /* CONFIG_GNRC_IPV6_NIB_QUEUE_PKT */
```

Reply:
Hm but it would be bad if packets cannot be queued even if there are 15 available slots because only one slot per neighbor is allowed. The current definition nevertheless looks a bit strange to me.

What about a neighbor is not allowed to have more than halve of the slots?

Reply:
I think this goes beyond my expertise, but I also think allowing for a fraction of the total slots is ok.

I also don't understand why `CONFIG_GNRC_IPV6_NIB_NUMOF` is used for size of the free slots.

Alternatively, we can make the number of slots >= `CONFIG_GNRC_IPV6_NIB_NUMOF * 4` and the queue size unlimited, and track the total number of used slots. Once 1/2 are used up, neighbors using more than (1/2 * total_num_of_slots / CONFIG_GNRC_IPV6_NIB_NUMOF) are not allowed to use any additional slots. This is a crude metric and detrimental if you only have a few on-link entries, but makes sure each entry will always receive at least 1 packet.

Reply:
> This is only happening during address resolution, right?
yes, only while in `INCOMPLETE`

Reply:
then I wonder why we need the per-node limit at all.
Sure we still need space in pktbuf to actually send out the neighbor solicitation (and receive the response), but that isn't deepening on the *number* of queued packets but their size - or do I misunderstand something here?

Reply:
@derMihai do you rather want to drop `CONFIG_GNRC_IPV6_NIB_QUEUE_PKT_CAP`?

Assume there is no CAP, what could happen?
At worst there is an edge case where the queue is not flushed.
Then one neighbor could take all the slots and at some point `_alloc_queue_entry()` will always fail.
Then we would notice the bug faster.

Reply:
> Sure we still need space in pktbuf to actually send out the neighbor solicitation (and receive the response), but that isn't deepening on the number of queued packets but their size - or do I misunderstand something here?

I think this is a different problem. What I'm trying to solve is the scenario where a neighbor gets flooded with packets and we run out of free queue entries. In that case, the neighbor(s) queue(s) will be filled with stale packets because we can't even `_alloc_queue_entry()` in order to drop the oldest packets. 

> Then one neighbor could take all the slots and at some point _alloc_queue_entry() will always fail.

AFAIK we should always drop the oldest, but that can't be done if we can't allocate in the first place.

The current capping solution doesn't guarantee this scenario won't happen, just makes it less probable.

Reply:
Also note that the default number of free queue entries is `CONFIG_GNRC_IPV6_NIB_NUMOF == 16`, which is rather small.

Reply:
You could as well drop the oldest packet, as soon as allocation fails and try to allocate again.
It could be that 16 slots are held by one neighbor and another neighbor for which allocation fails does not have an oldest packet to drop, so allocation fails again because one neighbor holds all the slots.

With your capacity limit per neighbor you want to make this case less likely, but also do not prevent this case.
Yes the idea is good, but at the same time you accept that a packet is not queued even though a slot could be allocated.

If I got that right I am not sure if this is really beneficial as long as we don´t note an issue that packets cannot be queued because one neighbor is taking most of the queue slots.
Did you observe this case?

Reply:
> If I got that right I am not sure if this is really beneficial as long as we don´t note an issue that packets cannot be queued because one neighbor is taking most of the queue slots.
Did you observe this case?

No, I only had issues with one host.

What bothers me most isn't failed allocations, but stale packets in a neighbor's queue. This goes against the first part of *be strict when sending and tolerant when receiving*.

> It could be that 16 slots are held by one neighbor and another neighbor for which allocation fails does not have an oldest packet to drop, so allocation fails again because one neighbor holds all the slots.

This is partially true. We have the `static _nib_onl_entry_t _nodes[CONFIG_GNRC_IPV6_NIB_NUMOF];
`. We could iterate through that and drop either from the first neighbor (faster) or the one with the largest queue (may be slower for large `CONFIG_GNRC_IPV6_NIB_NUMOF`)